mava/systems/ppo/sebulba/rec_ippo.py

```diff
@@ -0,0 +1,902 @@
+        params_source (ParamsSource): Source for fetching the latest network parameters
+        from the learner.
+        apply_fns (Tuple): Functions for running the actor and critic networks.
+        actor_device (Device): Actor device to use for rollout.
+        seeds (List[int]): Seeds for initializing the environment.
+        thread_lifetime (ThreadLifetime): Manages the thread's lifecycle.
+    """
+    name = threading.current_thread().name
+    print(f"{Fore.BLUE}{Style.BRIGHT}Thread {name} started{Style.RESET_ALL}")
+    actor_apply_fn, critic_apply_fn = apply_fns
+    num_agents, num_envs = config.system.num_agents, config.arch.num_envs
+    move_to_device = lambda x: jax.device_put(x, device=actor_device)
+
+    @jax.jit
+    def act_fn(
+        params: Params,
+        observation: Observation,
+        dones: chex.Array,
+        hstates: HiddenStates,
+        key: chex.PRNGKey,
+    ) -> Tuple:
+        """Get action and value."""
+
+        batched_observation = tree.map(lambda x: x[jnp.newaxis, :], observation)
+        ac_in = (
+            batched_observation,
+            dones[jnp.newaxis, :],
+        )
+        policy_hidden_state, actor_policy = actor_apply_fn(
+            params.actor_params, hstates.policy_hidden_state, ac_in
+        )
+        critic_hidden_state, value = critic_apply_fn(
+            params.critic_params, hstates.critic_hidden_state, ac_in
+        )
+
+        action = actor_policy.sample(seed=key)
+        log_prob = actor_policy.log_prob(action)
+
+        hstates = HiddenStates(policy_hidden_state, critic_hidden_state)
+        return action, log_prob, value, hstates
+
+    timestep = env.reset(seed=seeds)
+
+    # Initialise hidden states.
+    init_policy_hstate = ScannedRNN.initialize_carry(
+        (config.arch.num_envs, num_agents), config.network.hidden_state_dim
+    )
+    init_critic_hstate = ScannedRNN.initialize_carry(
+        (config.arch.num_envs, num_agents), config.network.hidden_state_dim
+    )
+    last_hstates = HiddenStates(init_policy_hstate, init_critic_hstate)
+    last_hstates = tree.map(move_to_device, last_hstates)
+
+    # Loop till the desired num_updates is reached.
+    while not thread_lifetime.should_stop():
+        # Rollout
+        traj: List[RNNPPOTransition] = []
+        episode_metrics: List[Dict] = []
+        actor_timings: Dict[str, List[float]] = defaultdict(list)
+        with RecordTimeTo(actor_timings["rollout_time"]):
+            for _ in range(config.system.rollout_length):
+                with RecordTimeTo(actor_timings["get_params_time"]):
+                    params = params_source.get()  # Get the latest parameters from the learner
+
+                last_obs = tree.map(move_to_device, timestep.observation)
+                last_dones = np.repeat(timestep.last(), num_agents).reshape(num_envs, -1)
+                last_dones = tree.map(move_to_device, last_dones)
```

Comment:
```suggestion
                last_obs = move_to_device(timestep.observation)
                last_dones = np.repeat(timestep.last(), num_agents).reshape(num_envs, -1)
                last_dones = move_to_device(last_dones)
```
I believe that jax's device_put works out of the box with pytrees.